File: LeetCode/CrackingTheCodeInterviewQuestions/Chapt3_Stack_Queue/min-stack.ts
/*
    Could have instead used vanilla object  {val: number, min: number} w/o creating class
    cant run using ts-node bc of MODULE_NOT_FOUND error

*/

class MinStackNode{
    val: number
    min: number
    constructor(val: number, min: number) {
        this.val = val;
        this.min = min;
    }
}

class MinStack {
    stack: Array<MinStackNode> // NOTE: Stack is array of tuples
    min: number

    constructor() {
        this.stack = []
    }

    

    push(val: number): void {
        this.stack.push(new MinStackNode(val, this.min))
        if (typeof(this.min) !== 'undefined') {
            if (val < this.min) 
            {
                this.min = val
            }
        } else {
            this.min = val
        }
    }

    pop(): void {
        if (this.stack.length) {
            const toBeDeleted = this.stack[this.stack.length - 1]
            if (this.min === toBeDeleted.val) 
            {
                this.min = toBeDeleted.min
            }
            this.stack.pop()
        }

    }

    top(): number {
        return this.stack.length ? this.stack[this.stack.length - 1].val : null
    }

    getMin(): number {
        return this.min ? this.min : null
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
//  const minStack = new MinStack()
//  minStack.push(-2)
//  minStack.push(0)
//  minStack.push(-3)
//  console.log(minStack.getMin())  // return -3
//  minStack.pop()
//  console.log(minStack.top()) // return 0# getting (<class 'TypeError'>, TypeError("'Node' object is not callable"), <traceback object at 0x1029b6988>) here
//  console.log(minStack.getMin())  // return -2

File: LeetCode/CrackingTheCodeInterviewQuestions/Chapt3_Stack_Queue/min_stack.py
'''
    Unresolved Issue: <class 'TypeError'>, TypeError("'Node' object is not callable"), <traceback object at 0x1029b6988> at the beginning of pop fxn being called

'''

class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.top = None
        self.min = None

    class Node:
        def __init__(self, val: int, next=None, prev_min=None):
            self.val = val
            self.next = next
            self.prev_min = prev_min

    def push(self, val: int) -> None:
        self.top = MinStack.Node(val, self.top, self.min)
        # update min if neccessary
        if self.min:  # NOTE: Need to check for None here
            if val < self.min:
                self.min = val
        else: # NOTE: Need to handle case min is None
            self.min = val

    def pop(self) -> None:
        if self.top:
            # update min if neccessary
            if self.min == self.top.val:  # if there are more than one nodes w/ same min this should not break
                self.min = self.top.prev_min
            self.top = self.top.next

    # getting (<class 'TypeError'>, TypeError("'Node' object is not callable"), <traceback object at 0x1029b6988>) here not sure why?
    def top(self):
        if self.top:
            return self.top.val
        else:
            return None  # choosing to return None when stack empty

    def getMin(self) -> int:
        return self.min


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

if __name__ == '__main__':
    minStack = MinStack()
    minStack.push(-2)
    minStack.push(0)
    minStack.push(-3)
    print(minStack.getMin())  # return -3
    minStack.pop()
    print(minStack.top())  # return 0# getting (<class 'TypeError'>, TypeError("'Node' object is not callable"), <traceback object at 0x1029b6988>) here
    print(minStack.getMin())  # return -2
